File: algorithms/2_medium/E/01609_even_odd_tree/bfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        curr_level = 0
        q = deque()
        q.append(root)

        while q:
            curr_len = len(q)
            prev_val = sys.maxsize if (curr_level & 1) == 1 else -sys.maxsize - 1

            for _ in range(curr_len):
                node = q.popleft()

                if (curr_level & 1) == (node.val & 1):
                    return False

                if (curr_level & 1) == 0:
                    if prev_val >= node.val:
                        return False
                else:
                    if prev_val <= node.val:
                        return False

                prev_val = node.val

                if node.left != None:
                    q.append(node.left)

                if node.right != None:
                    q.append(node.right)

            curr_level += 1

        return True

File: algorithms/2_medium/E/01609_even_odd_tree/bfs.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

impl Solution {
    pub fn is_even_odd_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let mut curr_level = 0_i32;
        let mut q: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        q.push_back(root.unwrap());

        while !q.is_empty() {
            let curr_len = q.len();
            let mut prev_val: Option<i32> = None;

            for _ in 0..curr_len {
                let node_rc = q.pop_front().unwrap();
                let node = node_rc.borrow();

                if (curr_level & 1) == (node.val & 1) {
                    return false;
                }
                
                if (curr_level & 1) == 0 {
                    if let Some(val) = prev_val && val >= node.val {
                        return false;
                    }
                } else {
                    if let Some(val) = prev_val && val <= node.val {
                        return false;
                    }
                }

                prev_val = Some(node.val);

                if let Some(left) = node.left.clone() {
                    q.push_back(left);
                }

                if let Some(right) = node.right.clone() {
                    q.push_back(right);
                }
            }

            curr_level += 1;
        }

        true
    }
}
